File: example/src/python/kv_cache/kv_client_example.py
# Copyright (c) Huawei Technologies Co., Ltd. 2023. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
State cache client python interface Test.
"""

from __future__ import absolute_import
import time

from datasystem.kv_client import KVClient


def assert_eq(value, expected_value):
    """Compare two values for equality."""
    if value != expected_value:
        raise RuntimeError(f"Assert failed, expect {expected_value}, but got {value}")


class KVClientExample:
    """This class shows the SDK usage example of the KVClient."""

    def __init__(self, host, port):
        self._host = host
        self._port = port

    def set_data_example(self):
        """This function shows the basic usage of set/get/del."""
        client = KVClient(self._host, self._port)
        client.init()
        key = "key"
        expected_val = "value"
        client.set(key, expected_val)

        val = client.get([key], True)
        assert_eq(val[0], expected_val)

        client.delete([key])

    def set_value_with_generate_key_example(self):
        """This function shows the basic usage of set_value/generate_key."""
        client = KVClient(self._host, self._port)
        client.init()
        expected_val = "value"
        key = client.set_value(expected_val)

        val = client.get([key], True)
        assert_eq(val[0], expected_val)

        client.delete(["key"])

        key = client.generate_key()
        client.set(key, expected_val)
        client.delete([key])

    def set_data_with_ttl_example(self):
        """This function shows the basic usage of TTL(time to live).
        This value is valid only for the TTL time and will be deleted after the time expires."""
        client = KVClient(self._host, self._port)
        client.init()
        key = "key"
        expected_val = "value"
        client.set(key, expected_val, ttl_second=2)

        val = client.get([key])
        decode_val = val[0].decode()
        if decode_val != expected_val:
            raise RuntimeError(f"[set_data_with_ttl] Get value failed, expect {expected_val}, but got {val}")

        time.sleep(3)
        try:
            client.get([key])
        except RuntimeError as e:
            if "Cannot get objects from worker" not in str(e):
                raise RuntimeError('get failed ') from e

    def set_data_with_l2_cache_example(self):
        """This function shows the basic usage of write_mode. If the L2Cache is configured on the server,
        data can be written to the L2Cache to improve data reliability."""
        client = KVClient(self._host, self._port)
        client.init()
        expected_val = "value"
        client.set("key", expected_val)


if __name__ == '__main__':
    example = KVClientExample('127.0.0.1', 31501)
    example.set_data_example()
    example.set_value_with_generate_key_example()
    example.set_data_with_ttl_example()
    example.set_data_with_l2_cache_example()

File: example/java/src/main/java/org/yuanrong/datasystem/example/KVClientExample.java
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yuanrong.datasystem.example;

import org.yuanrong.datasystem.ConnectOptions;
import org.yuanrong.datasystem.kv.KVClient;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.logging.Logger;

/**
 * The example of using KVClient.
 *
 * @since 2022-09-01
 */
public class KVClientExample {
    private static final Logger log = Logger.getLogger(KVClientExample.class.toString());

    /**
     * Print ByteBuffer.
     *
     * @param buffer The ByteBuffer.
     */
    public static void printByteBuffer(ByteBuffer buffer) {
        byte[] bytes = new byte[buffer.remaining()];
        buffer.get(bytes);
        String value = new String(bytes, StandardCharsets.UTF_8);
        String message = "The value is : " + value;
        log.info(message);
    }

    public static void main(String[] args) {
        log.info("---- Start the KV java api demo");
        String ip = args[0];
        int port = Integer.parseInt(args[1]);
        ConnectOptions connectOpts = new ConnectOptions(ip, port);
        KVClient client = new KVClient(connectOpts);
        try {
            List<String> keys = Arrays.asList("key1", "key2", "key3");
            String tmp = "abc";
            byte[] bytes = tmp.getBytes(StandardCharsets.UTF_8);
            ByteBuffer value = ByteBuffer.allocateDirect(bytes.length);
            value.put(bytes);
            // client set and get
            for (int i = 0; i < 3; i++) {
                client.set(keys.get(i), value);
            }
            ByteBuffer getValue = client.get(keys.get(0));
            List<ByteBuffer> getValues = client.get(keys);
            client.del(keys);
            String message = "The size of value list is : " + getValues.size();
            log.info(message);
            printByteBuffer(getValue);
        } finally {
            client.close();
        }
    }
}